src/plots/mapbox/mapbox.js

```diff
@@ -316,7 +316,12 @@ proto.updateData = function(calcData) {
         traceObj = traceHash[trace.uid];
 
         if(traceObj) {
-            traceObj.update(calcTrace);
+            if(traceObj.type === trace.type) {
+                traceObj.update(calcTrace);
+            } else {
+                traceObj.dispose();
+                traceHash[trace.uid] = trace._module.plot(this, calcTrace);
+            }
         } else if(trace._module) {
             traceHash[trace.uid] = trace._module.plot(this, calcTrace);
         }
```

Comment:
What about writing this whole logic like this?
```js
var updated;
if(traceObj) {
    if(traceObj.type === trace.type) {
        traceObj.update(calcTrace);
        updated = true;
    } else {
        traceObj.dispose();
    }
}
if(!updated && trace._module) {
    traceHash[trace.uid] = trace._module.plot(this, calcTrace);
}
```

Reply:
Ok sure - done in https://github.com/plotly/plotly.js/pull/4295/commits/5427dc047bb3e2e58333b1662b8e74d1f20f7557